### PR TITLE
[SDAG] [X86] Extend SplitVecOp_VSETCC for STRICT_FSETCCS

### LINKED ISSUE #116767: LLVM ERROR: Do not know how to split this operator's operand
Compiler Explorer: https://godbolt.org/z/s5rzrox6M
```
LLVM ERROR: Do not know how to split this operator's operand!

PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /opt/compiler-explorer/clang-trunk/bin/llc -o /app/output.s -x86-asm-syntax=intel -mtriple=x86_64 -mcpu=skx <source>
1.	Running pass 'Function Pass Manager' on module '<source>'.
2.	Running pass 'X86 DAG->DAG Instruction Selection' on function '@foo'
 #0 0x000000000395b818 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/opt/compiler-explorer/clang-trunk/bin/llc+0x395b818)
 #1 0x00000000039593dc SignalHandler(int) Signals.cpp:0:0
 #2 0x00007b0ec3442520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #3 0x00007b0ec34969fc pthread_kill (/lib/x86_64-linux-gnu/libc.so.6+0x969fc)
 #4 0x00007b0ec3442476 gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x42476)
 #5 0x00007b0ec34287f3 abort (/lib/x86_64-linux-gnu/libc.so.6+0x287f3)
 #6 0x000000000075cfb6 llvm::UniqueStringSaver::save(llvm::Twine const&) (.cold) StringSaver.cpp:0:0
 #7 0x00000000038b1938 (/opt/compiler-explorer/clang-trunk/bin/llc+0x38b1938)
 #8 0x00000000037dc806 llvm::DAGTypeLegalizer::SplitVectorOperand(llvm::SDNode*, unsigned int) (/opt/compiler-explorer/clang-trunk/bin/llc+0x37dc806)
 #9 0x00000000037afb7d llvm::DAGTypeLegalizer::run() (/opt/compiler-explorer/clang-trunk/bin/llc+0x37afb7d)
#10 0x00000000037b0041 llvm::SelectionDAG::LegalizeTypes() (/opt/compiler-explorer/clang-trunk/bin/llc+0x37b0041)
#11 0x0000000003718ff2 llvm::SelectionDAGISel::CodeGenAndEmitDAG() (/opt/compiler-explorer/clang-trunk/bin/llc+0x3718ff2)
#12 0x000000000371bcac llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) (/opt/compiler-explorer/clang-trunk/bin/llc+0x371bcac)
#13 0x000000000371d90f llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) (/opt/compiler-explorer/clang-trunk/bin/llc+0x371d90f)
#14 0x0000000003709ce1 llvm::SelectionDAGISelLegacy::runOnMachineFunction(llvm::MachineFunction&) (/opt/compiler-explorer/clang-trunk/bin/llc+0x3709ce1)
#15 0x00000000029e60db llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (.part.0) MachineFunctionPass.cpp:0:0
#16 0x0000000002f3d912 llvm::FPPassManager::runOnFunction(llvm::Function&) (/opt/compiler-explorer/clang-trunk/bin/llc+0x2f3d912)
#17 0x0000000002f3dba1 llvm::FPPassManager::runOnModule(llvm::Module&) (/opt/compiler-explorer/clang-trunk/bin/llc+0x2f3dba1)
#18 0x0000000002f3f546 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/opt/compiler-explorer/clang-trunk/bin/llc+0x2f3f546)
#19 0x0000000000881e09 compileModule(char**, llvm::LLVMContext&) llc.cpp:0:0
#20 0x0000000000771476 main (/opt/compiler-explorer/clang-trunk/bin/llc+0x771476)
#21 0x00007b0ec3429d90 (/lib/x86_64-linux-gnu/libc.so.6+0x29d90)
#22 0x00007b0ec3429e40 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e40)
#23 0x0000000000879b1e _start (/opt/compiler-explorer/clang-trunk/bin/llc+0x879b1e)
Program terminated with signal: SIGSEGV
Compiler returned: 139
```

### PR DESCRIPTION
**Pull request by abhishek-kaushik22**
Closes #116767

### CODE REVIEW

**Comment by abhishek-kaushik22** [comment]
@MalaySanghi @phoebewang @e-kud Can you please review?

**Review by MalaySanghi — APPROVED** [review]
LGTM

File: llvm/test/CodeGen/X86/vec-strict-cmp-512-skx-fp64.ll [hunk 1]
```diff
@@ -0,0 +1,31 @@
+; RUN: llc < %s -mtriple=x86_64 -mcpu=skx | FileCheck %s --check-prefixes=SKX
+
+define <16 x i32> @test_v16f64_ogt(<16 x i32> %a, <16 x i32> %b, <16 x double> %f1, <16 x double> %f2) #0 {
```
[the remark this answers is not in the document]
Comment (abhishek-kaushik22) [reply llvm/test/CodeGen/X86/vec-strict-cmp-512-skx-fp64.ll:3]: I made the test similar to the existing test which have both the prefix and the attribute

https://github.com/llvm/llvm-project/blob/main/llvm/test/CodeGen/X86/vec-strict-cmp-512-skx.ll#L40

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp [hunk 2]
```diff
@@ -4236,7 +4237,8 @@ SDValue DAGTypeLegalizer::SplitVecOp_TruncateHelper(SDNode *N) {
 }
 
 SDValue DAGTypeLegalizer::SplitVecOp_VSETCC(SDNode *N) {
-  bool isStrict = N->getOpcode() == ISD::STRICT_FSETCC;
+  bool isStrict = N->getOpcode() == ISD::STRICT_FSETCC ||
+                  N->getOpcode() == ISD::STRICT_FSETCCS;
```
Comment (phoebewang) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:4241]: There're many calls to `N->getOpcode()`. Define a variable for it.

**Review by phoebewang — APPROVED** [review]
LGTM.

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp [hunk 3]
```diff
@@ -4252,21 +4254,19 @@ SDValue DAGTypeLegalizer::SplitVecOp_VSETCC(SDNode *N) {
...
   EVT WideResVT = EVT::getVectorVT(Context, MVT::i1, PartEltCnt*2);
 
-  if (N->getOpcode() == ISD::SETCC) {
+  if (Opc == ISD::SETCC) {
     LoRes = DAG.getNode(ISD::SETCC, DL, PartResVT, Lo0, Lo1, N->getOperand(2));
     HiRes = DAG.getNode(ISD::SETCC, DL, PartResVT, Hi0, Hi1, N->getOperand(2));
-  } else if (N->getOpcode() == ISD::STRICT_FSETCC) {
-    LoRes = DAG.getNode(ISD::STRICT_FSETCC, DL,
-                        DAG.getVTList(PartResVT, N->getValueType(1)),
+  } else if (Opc == ISD::STRICT_FSETCC || Opc == ISD::STRICT_FSETCCS) {
```
Comment (phoebewang) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:4260]: `else if (isStrict)`

**Review by e-kud — APPROVED** [review]
LGTM. However I don't see any successful build. Maybe a merge from main is required?

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp [hunk 4]
```diff
@@ -4236,7 +4237,8 @@ SDValue DAGTypeLegalizer::SplitVecOp_TruncateHelper(SDNode *N) {
 }
 
 SDValue DAGTypeLegalizer::SplitVecOp_VSETCC(SDNode *N) {
-  bool isStrict = N->getOpcode() == ISD::STRICT_FSETCC;
+  auto Opc = N->getOpcode();
```
Comment (RKSimon) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:4240]: (style) don't use auto - `unsigned Opc`

**Review by topperc — APPROVED** [review]
LGTM

**Comment by abhishek-kaushik22** [comment]
Can someone please merge?

**Pull request merged**